shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt

```diff
@@ -0,0 +1,63 @@
+package com.ivy.domain.usecase
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+
+class BalanceBuilder {
+
+    private val balance = mutableMapOf<AssetCode, NonZeroDouble>()
+
+    fun processDeposit(
+        incomes: Map<AssetCode, PositiveDouble>,
+        transferIn: Map<AssetCode, PositiveDouble>,
+    ) {
+        getAmount(incomes, transferIn).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + value.value)
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    fun processWithdrawal(
+        expenses: Map<AssetCode, PositiveDouble>,
+        transferOut: Map<AssetCode, PositiveDouble>,
+    ) {
+        getAmount(expenses, transferOut).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + (-value.value))
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    private fun getAmount(
+        firstMap: Map<AssetCode, PositiveDouble>,
+        secondMap: Map<AssetCode, PositiveDouble>,
+    ): Map<AssetCode, PositiveDouble> {
+        val amount = firstMap.toMutableMap()
+        combine(amount, secondMap)
+        return amount
+    }
+
+    private fun combine(
+        amount: MutableMap<AssetCode, PositiveDouble>,
+        secondMap: Map<AssetCode, PositiveDouble>
+    ) {
+        secondMap.forEach { (asset, value) ->
+            amount.merge(asset, value) { firstMap, secondMap ->
```

Comment:
This namings are confusing. Let's call the two maps `a` and `b` or something consistent. Now `firstMap` and `amount` are clashing and I can't easily understand the code